applications/launchpad/backend/src/docker/workspace.rs

```diff
@@ -289,11 +289,11 @@ impl TariWorkspace {
         Ok(ids)
     }
 
-    /// Create and return a [`Stream`] of [`LogMessage`] instances for the `name`d container in the workspace.  
+    /// Create and return a [`Stream`] of [`LogMessage`] instances for the `name`d container in the workspace.
     pub fn logs(
         &self,
         container_name: &str,
-        docker: Docker,
+        docker: &Docker,
```

Comment:
How about to put `Docker` into the `TariWorkspace`. We have to pass a reference to some methods calls, but it will me more convenient if we could keep the reference with the workspace.

1-st case, if you can put the `Docker` inside the `struct`:

```rust
pub struct TariWorkspace {
    name: String,
    config: LaunchpadConfig,
    containers: HashMap<String, ContainerState>,

    // Extra field
    docker: Docker,
}
```

It's the preferred way, but if a `Docker` instance aren't available at the moment of creation `TariWorkspace` you can add wrap it with an `Option`:

2-nd case:

```rust
pub struct TariWorkspace {
    name: String,
    config: LaunchpadConfig,
    containers: HashMap<String, ContainerState>,

    // Extra field
    docker: Option<Docker>,
}

impl TariWorkspace {

    // Methods to assign and access docker (not necessary if `Docker` is not optional)
    fn assign_docker(&mut self, docker: Docker) {
        self.docker = Some(docker);
    }

    fn docker_ref(&self) -> Result<&Docker, Error> {
        // Or `DockerWrapperError`
        self.docker.as_ref().ok_or_else(|| Error::msg("Docker not set to the TariWorkspace"))
    }
    
    fn docker(&self) -> Result<Docker, Error> {
        // Or `DockerWrapperError`
        self.docker.clone().ok_or_else(|| Error::msg("Docker not set to the TariWorkspace"))
    }
    
    pub fn logs(
        &self,
        container_name: &str,
    ) -> Option<impl Stream<Item = Result<LogMessage, DockerWrapperError>>> {

        // It returns `None` here if a docker reference is not available.
        let docker = self.docker_ref().ok()?;
        // Or use `Option` directly here (than `docker_ref` method is not necessary):
        // let docker = self.docker.as_ref()?;

        let options = LogsOptions::<String> {
            follow: true,
            stdout: true,
            stderr: true,
            ..Default::default()
        };
        self.containers.get(container_name).map(move |container| {
            let id = container.id();
            docker
                .logs(id.as_str(), Some(options))
                .map(|log| log.map(LogMessage::from).map_err(DockerWrapperError::from))
        })
    }    
    
    pub async fn start_service(
        &mut self,
        image: ImageType,
        registry: Option<String>,
        tag: Option<String>,
    ) -> Result<String, DockerWrapperError> {

    	let docker = self.docker()?;
    	// Use docker here

    }
}
```

If we can do that, we can eliminate places where we have to pass `Docker` to methods directly.

Reply:
Anyway, we can put that idea to a todo list as a variant of code layout. I keep doing review of other parts...

Reply:
Thanks for the suggestion, probably good for another PR. This PR tried to avoid refactoring and focuses on fixing clippy errors/minor corrections in the least lines of code